Ldap Group Provider

eformat opened the pull request:
Reopening work previously closed  - this PR supersedes #8335 #10116

Adds
- ldap group provider

cc: @bitsondatadev 

Please advise on what you think needs finishing off. I have tested this against a local FreeIPA ldap setup OK and have joined your trino slack dev channel as well. cheers.

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:
```diff
@@ -0,0 +1,198 @@
+        implements GroupProvider
+{
+    private static final Logger log = Logger.get(LdapGroupProvider.class);
+    private static final Pattern EXTRACT_GROUP_PATTERN = Pattern.compile("(?i)cn=([^,]*),(.*)$");
```
Instead of using PatternMatcher can we try something like this 
```    private String getRelativeDistinguishedName(String distinguishedName)
            throws NamingException {
        LdapName ldn = new LdapName(distinguishedName);
        List<Rdn> rdns = ldn.getRdns();
        if (rdns.isEmpty()) {
            throw new NamingException("DN is empty");
        }
        Rdn rdn = rdns.get(rdns.size()-1);
        if (rdn.getType().equalsIgnoreCase(groupNameAttr)) {
            String groupName = (String)rdn.getValue();
            return groupName;
        }
        throw new NamingException("Unable to find RDN: The DN " +
                distinguishedName + " is malformed.");
    }

eformat replied:
sure, modifying ...

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:
```diff
@@ -0,0 +1,198 @@
+    private final List<String> userBindSearchPatterns;
+    private final Optional<String> userBaseDistinguishedName;
+    private final String groupBaseDistinguishedName;
+    private String groupSearchFilter;
```
Any reason for non-final variable ?

eformat replied:
ha no ;) fixing.

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:
```diff
@@ -0,0 +1,198 @@
+                });
+    }
+
+    String lookupUserDistinguishedName(String userBaseDistinguishedName, String groupAuthorizationSearchPattern,
```
Can we move each argument to a new line and provide a access sepcifier to this

eformat replied:
yup, fixing.

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:
```diff
@@ -0,0 +1,198 @@
+                        log.debug(e, "Unknown exception, %s", e.getMessage());
+                        return "";
+                    }
+                }).filter(Objects::nonNull).collect(Collectors.toSet());
```
Can we use `guava` Immutable collections.

eformat replied:
core java17 now has Collectors.toUnmodifiableSet() which achieves the same thing?, will put this in here

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:
```diff
@@ -0,0 +1,198 @@
+        }
+    }
+
+    private Set<String> getGroupsWithBindDistinguishedName(String user)
```
How about AD approach i.e if a user has `memberOf` attribute as a part of its `LdapObject`.

eformat replied:
OK .. AFAIK there are 3 types  of group membership: 

RFC 2307 (posix) - from 1998, venerable :) - in this draft PR so far
RFC2307bis - Active Directory - rfc never made it out of draft, despite Microsoft using it
Augmented Active Directory.

let me investigate this, see if we can easily add in AD support ...

eformat replied:
this is a good representation

![image](https://github.com/trinodb/trino/assets/712608/830febed-1c20-4b01-a033-fc7ddb783647)

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProviderConfig.java:
```diff
@@ -0,0 +1,173 @@
+    private String groupSearchFilter;
+    private Duration ldapCacheTtl = new Duration(1, TimeUnit.HOURS);
+
+    @NotNull
```
Can we apply additional assertion on `NotEmpty`

eformat replied:
yup, adding ..

Praveen2112 commented on plugin/trino-password-authenticators/src/test/java/io/trino/plugin/password/ldap/LdapUtil.java:
```diff
@@ -36,7 +36,8 @@
 
     private LdapUtil()
-    {}
+    {
```
Can we revert this change.

eformat replied:
yes, i will tidy up the formatting in this class in general actually.

Praveen2112 commented on plugin/trino-password-authenticators/src/test/java/io/trino/plugin/password/ldap/TestLdapUtil.java:
```diff
@@ -0,0 +1,54 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.password.ldap;
```
Rearranging the code can land in a separate commit.

eformat replied:
can you clarify please = do you mean put this test class in a separate commit ?

eformat replied:
oh .. this is a duplicate test i see now, will remove

Praveen2112 reviewed (commented):
Thanks for working on this. It would be nice if we could have a PTs for this.

eformat commented:
> PTs

ah yup .. Product Tests .. also i will write some Doco ;) for this .. WIP.

eformat commented:
BTW i have filled in the CLA and emailed it back am ignoring the bot ;)

eformat commented:
hi @Praveen2112 PTAL, i have added in support for these three common ldap schema use cases now.
this of course needs proper documentation. the test container in the code uses the scheme in (1), will need to find some schemas that can mimic (2),(3) for real life testing (unit tests cover the config though).

(1) RFC2307 (posix), RFC2307bis

```
  dn: ou=users,dc=example,dc=com
  objectClass: organizationalUnit
  ou: users

  dn: cn=Jane,ou=users,dc=example,dc=com
  objectClass: person
  objectClass: organizationalPerson
  objectClass: inetOrgPerson
  cn: Jane
  sn: Smith
  displayName: Jane Smith
  mail: jane.smith@example.com

  dn: cn=Jim,ou=users,dc=example,dc=com
  objectClass: person
  objectClass: organizationalPerson
  objectClass: inetOrgPerson
  cn: Jim
  sn: Adams
  displayName: Jim Adams
  mail: jim.adams@example.com

  dn: ou=groups,dc=example,dc=com
  objectClass: organizationalUnit
  ou: groups

  dn: cn=admins,ou=groups,dc=example,dc=com 
  objectClass: groupOfNames
  cn: admins
  owner: cn=admin,dc=example,dc=com
  description: System Administrators
  member: cn=Jane,ou=users,dc=example,dc=com 
  member: cn=Jim,ou=users,dc=example,dc=com
```

The group is a first-class entry in the LDAP server (e.g. "cn=admins,ou=groups,dc=example,dc=com")
Members of a group are listed with an identifying reference as attributes on the group (e.g. "member")

(2) Active Directory

```
dn: ou=users,dc=example,dc=com
objectClass: organizationalUnit
ou: users

dn: cn=Jane,ou=users,dc=example,dc=com
objectClass: person
objectClass: organizationalPerson
objectClass: inetOrgPerson
objectClass: testPerson
cn: Jane
sn: Smith
displayName: Jane Smith
mail: jane.smith@example.com
memberOf: admins 

dn: cn=Jim,ou=users,dc=example,dc=com
objectClass: person
objectClass: organizationalPerson
objectClass: inetOrgPerson
objectClass: testPerson
cn: Jim
sn: Adams
displayName: Jim Adams
mail: jim.adams@example.com
memberOf: admins
```

The user’s group memberships are listed as attributes on the user, and the group does not exist as an entry on the server. e.g "memberOf"

(3) Augmented Active Directory

```
dn: ou=users,dc=example,dc=com
objectClass: organizationalUnit
ou: users

dn: cn=Jane,ou=users,dc=example,dc=com
objectClass: person
objectClass: organizationalPerson
objectClass: inetOrgPerson
objectClass: testPerson
cn: Jane
sn: Smith
displayName: Jane Smith
mail: jane.smith@example.com
memberOf: cn=admins,ou=groups,dc=example,dc=com 

dn: cn=Jim,ou=users,dc=example,dc=com
objectClass: person
objectClass: organizationalPerson
objectClass: inetOrgPerson
objectClass: testPerson
cn: Jim
sn: Adams
displayName: Jim Adams
mail: jim.adams@example.com
memberOf: cn=admins,ou=groups,dc=example,dc=com

dn: ou=groups,dc=example,dc=com
objectClass: organizationalUnit
ou: groups

dn: cn=admins,ou=groups,dc=example,dc=com 
objectClass: groupOfNames
cn: admins
owner: cn=admin,dc=example,dc=com
description: System Administrators
member: cn=Jane,ou=users,dc=example,dc=com
member: cn=Jim,ou=users,dc=example,dc=com
```

The user’s group memberships are listed as attributes on the user. e.g. "memberOf: cn=admins,ou=groups,dc=example,dc=com"

The group is a first-class entry on the LDAP server. e.g. "dn: cn=admins,ou=groups,dc=example,dc=com"

eformat commented:
Commit https://github.com/trinodb/trino/pull/17518/commits/0d0c56aa3d362cf4d5fd59c8ae5aef9f7fa63d1a

Added testing support for active directory schema based testing.
Added testing support for multii-value attributes e.g memberOf

eformat commented:
Documentation for LDAP Group provider now added in this [commit](https://github.com/trinodb/trino/pull/17518/commits/991cea15512a7cf604ccf657384aa97ee4096e3a).

eformat commented:
CLA - i signed this and returned it on 16/5/23 cheers.

eformat commented:
ping @Praveen2112 .. got time to review this ?

Praveen2112 commented:
Will be reviewing this.

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java:
```diff
@@ -65,7 +65,8 @@ public LdapAuthenticator(LdapAuthenticatorClient client, LdapAuthenticatorConfig
         this.userBindSearchPatterns = ldapAuthenticatorConfig.getUserBindSearchPatterns();
-        this.groupAuthorizationSearchPattern = Optional.ofNullable(ldapAuthenticatorConfig.getGroupAuthorizationSearchPattern());
+        this.groupAuthorizationSearchPattern =
+                Optional.ofNullable(ldapAuthenticatorConfig.getGroupAuthorizationSearchPattern());
```
can we revert this change ?

eformat replied:
we could .. but this AuthorizationSearchPattern is not used on LDAP Group Provider.
it was a copy and paste from the LDAP Auth Provider.
so i would suggest no need to revert as it was not used ?

Praveen2112 replied:
Sorry I meant moving `Optional.ofNullable(ldapAuthenticatorConfig.getGroupAuthorizationSearchPattern());` to a new line

Praveen2112 replied:
Or we could have a dedicated commit for the formatting changes.

Praveen2112 commented:
Can we squash the commit into one ?

Praveen2112 commented:
Can squash the commit to logical ones ?

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:
```diff
@@ -0,0 +1,324 @@
+            throw new LdapGroupMembershipException("both ldap.group-membership-attribute and "
+                    + "ldap.group-user-membership-attribute cannot be empty");
+        }
+        if (this.groupMembershipAttribute.isPresent() && this.groupNameAttribute.isEmpty()) {
```
Can we move this verification to the config class ?

eformat replied:
We could - but its analogous to the FileProvider - where validation handled in the constructor for the Provider instance itself. The Config classes have no such constructors.

Praveen2112 replied:
We don't need to have a constructor for validation. We could annotate a method with `AssertTrue` or `AssertFalse` to apply corresponding validation. 

For example 
```
    @AssertFalse(message = "Connecting to the LDAP server without SSL enabled requires `ldap.allow-insecure=true`")
    public boolean validateGroupMembership()
    {
        return this.groupMembershipAttribute.isPresent() && this.groupUserMembershipAttribute.isPresent()
    }

Something like this

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:
```diff
@@ -0,0 +1,324 @@
+        return getOnlyElement(userDistinguishedNames);
+    }
+
+    static String extractGroupName(
```
Can we move the static methods to a Util class ?

eformat replied:
We could - but its analogous to the FileGroupProvider - which has static methods in it.

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:
```diff
@@ -0,0 +1,324 @@
+                log.debug("Cannot extract group [%s] [%s] [%s]", groupNameAttribute, groupUserMembershipAttribute, source);
+            }
+            return null;
+        }).filter(Objects::nonNull).collect(Collectors.toUnmodifiableSet());
```
Any specific for not using guava `ImmutableSet` 
Ref : https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#prefer-guava-immutable-collections

eformat replied:
yeah, i did see this.  Is there a reason that core Java Collectors.toUnmodifiableSet() is not OK here? We do not require ordered Sets - and if we did, would use a different container to explicitly make that behavior so. I can change it but i can see several other uses if this in the code base outside ldap.

Praveen2112 replied:
The main reason is deterministic iteration which could help us during debugging and to have consistency across the code base.  


> i can see several other uses of this in the code base outside ldap.

The usage is specific to iceberg connector I guess - other part of the codebase still use `ImmutableSet` instead of `UnmodifiableSet`.

Praveen2112 replied:
`.filter` and `.collect` can be moved to a new line

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:
```diff
@@ -0,0 +1,324 @@
+        try {
+            for (String userBindSearchPattern : userBindSearchPatterns) {
+                String userSearchFilter = replaceUser(userBindSearchPattern, user);
+                if (this.groupUserMembershipAttribute.isPresent()) {
```
Is it possible to have an interface for the implementation and based on the config we can inject them during the runtime ? i.e based on `groupUserMembershipAttribute` or not.

eformat replied:
yeah good idea .. this is a bit clunky .. let me investigate ..

eformat replied:
resolved with https://github.com/trinodb/trino/pull/17518/commits/4e4c8fe3d8503bada7f94ae026820d6ab03e63d0
PTAL

eformat commented:
> Can squash the commit to logical ones ?

yes, done

eformat commented:
@Praveen2112 .. thanks for reviewing this, see my replies. I think the for the remaining unresolved comments - i could make more changes, but am unsure what value they would bring.

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java:
```diff
@@ -187,22 +192,24 @@ private String lookupUserDistinguishedName(String user)
     }
 
-    private static String replaceUser(String pattern, String user)
+    static String replaceUser(String pattern, String user)
```
Or we could have a Util class for this

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java:
```diff
@@ -187,22 +192,24 @@ private String lookupUserDistinguishedName(String user)
         }
         if (userDistinguishedNames.size() > 1) {
-            String message = format("Multiple group membership results for user [%s]: %s", user, userDistinguishedNames);
+            String message = format("Multiple group membership results for user [%s]: %s", user,
```
We can revert his formatting changes

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java:
```diff
@@ -125,7 +127,8 @@ private Principal authenticateWithUserBind(Credential credential)
                     String searchBase = userBaseDistinguishedName.orElseThrow();
                     String groupSearch = replaceUser(groupAuthorizationSearchPattern.get(), user);
-                    if (!client.isGroupMember(searchBase, groupSearch, userDistinguishedName, credential.getPassword())) {
+                    if (!client.isGroupMember(searchBase, groupSearch, userDistinguishedName,
```
Can we revert the formatting changes ?

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java:
```diff
@@ -187,22 +192,24 @@ private String lookupUserDistinguishedName(String user)
         String searchBase = userBaseDistinguishedName.orElseThrow();
         String searchFilter = replaceUser(groupAuthorizationSearchPattern.orElseThrow(), user);
-        Set<String> userDistinguishedNames = client.lookupUserDistinguishedNames(searchBase, searchFilter, bindDistinguishedName.orElseThrow(), bindPassword.orElseThrow());
+        Set<String> userDistinguishedNames = client.lookupUserDistinguishedNames(searchBase, searchFilter,
```
Something like this 
```
        Set<String> userDistinguishedNames = client.lookupUserDistinguishedNames(
                searchBase, 
                searchFilter, 
                bindDistinguishedName.orElseThrow(), 
                bindPassword.orElseThrow());
                
ditto.

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:
```diff
@@ -0,0 +1,324 @@
+import io.trino.spi.security.AccessDeniedException;
+import io.trino.spi.security.GroupProvider;
+
+import javax.inject.Inject;
```
We have migrated from `javax.inject.Inject` to a guice based `Inject`. Can we rebase on the latest master to ensure that all checks are applied and the build is green ?

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:
```diff
@@ -0,0 +1,324 @@
+    private final LoadingCache<String, Set<String>> groupCache;
+
+    @Inject
+    public LdapGroupProvider(LdapClient client, LdapGroupProviderConfig ldapConfig)
```
Can we rename `ldapConfig` to `ldapGroupProviderConfig` ?

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:
```diff
@@ -0,0 +1,324 @@
+
+    private Set<String> getGroupsWithBindDistinguishedName(String user)
+    {
+        Set<String> groupSet = Collections.emptySet();
```
How about `ImmutableSet.of` for consistency across the code base ?

Praveen2112 commented on plugin/trino-password-authenticators/src/test/java/io/trino/plugin/password/ldap/DisposableSubContext.java:
```diff
@@ -0,0 +1,46 @@
+/*
```
This change can be extracted as separate commit

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProviderQuery.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.util.Set;
+
+public interface LdapGroupProviderQuery
```
We could rename the interface to `GroupProvider`

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:
```diff
@@ -93,6 +94,18 @@ public LdapGroupProvider(LdapClient client, LdapGroupProviderConfig ldapConfig)
                 CacheBuilder.newBuilder()
                         .expireAfterWrite(ldapConfig.getLdapCacheTtl().toMillis(), MILLISECONDS),
                 CacheLoader.from(this::getGroupsWithBindDistinguishedName));
+        if (this.groupUserMembershipAttribute.isPresent()) {
+            this.ldapGroupProviderQueryUser = Optional.of(new LdapGroupProviderQueryUser());
+        }
+        else {
+            this.ldapGroupProviderQueryUser = Optional.empty();
+        }
+        if (this.groupMembershipAttribute.isPresent()) {
+            this.ldapGroupProviderQueryGroup = Optional.of(new LdapGroupProviderQueryGroup());
+        }
+        else {
+            this.ldapGroupProviderQueryGroup = Optional.empty();
+        }
```
We can inject `LdapGroupProviderQuery` implementation based on the config property so we don't have to maintain two variables one for `ldapGroupProviderQueryUser` and `ldapGroupProviderQueryGroup`

Praveen2112 replied:
Please check the usage of `AbstractConfigurationAwareModule` for specifying which implementation to bind.

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProviderQueryUser.java:
```diff
@@ -0,0 +1,70 @@
+        try {
+            if (ldapGroupProvider.groupUserMembershipAttribute.isPresent()) {
+                try {
+                    groupSet = ldapGroupProvider.lookupUserAttributes(
```
Instead of using `LdapGroupProvider` can we inject `LdapClient` use and invoke the corresponding ldap query ?

Praveen2112 commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProviderQueryGroup.java:
```diff
@@ -0,0 +1,78 @@
+    {
+        Set<String> groupSet = Collections.emptySet();
+        try {
+            String userDistinguishedName = ldapGroupProvider.lookupUserDistinguishedName(
```
Instead of using `LdapGroupProvider` can we inject `LdapClient` and invoke corresponding LdapQuery ?

eformat commented:
> Can we resolve the conflict related to `security.rst` to ensure the CI is working.

hi .. yes, can do .. its a one-liner for adding in a document entry top-level

eformat commented:
looks like the original doc -> moved to markdown in master. i will need to add it back in with a separate change i think, will be easier:

security/group-ldap

eformat commented:
@Praveen2112 - also, please note the new test ldap image needed for microsoft schema

https://github.com/trinodb/docker-images/pull/166

this will be a prerequisite for the tests to complete.

eformat commented:
rebasing ...

eformat commented:
tests will run OK once ldap test image merges

https://github.com/trinodb/docker-images/pull/166

```
[WARNING] Tests run: 53, Failures: 0, Errors: 0, Skipped: 4, Time elapsed: 29.619 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[WARNING] Tests run: 53, Failures: 0, Errors: 0, Skipped: 4
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  45.123 s
[INFO] Finished at: 2023-06-13T14:17:01+10:00
```

ksobolew commented on plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:
```diff
@@ -0,0 +1,284 @@
+            throw new LdapGroupMembershipException("ldap.group-membership-attribute is set "
+                    + "ldap.group-name-attribute cannot be empty");
+        }
+        this.groupCache = buildNonEvictableCacheWithWeakInvalidateAll(
+                CacheBuilder.newBuilder()
+                        .expireAfterWrite(ldapConfig.getLdapCacheTtl().toMillis(), MILLISECONDS),
+                CacheLoader.from(this::getGroupsWithBindDistinguishedName));
```
Note: I'm intending to add a generic caching layer in Trino for group providers. See #18174. I'll take some inspiration from how you did this here :) If that caching layer is accepted, caching here would be unnecessary.

kokosing commented:
Consider using https://github.com/trinodb/trino/pull/18174, CC: @ksobolew

Praveen2112 commented:
Can you rebase the PR. It looks like there were a few conflicts to be resolved

eformat commented:
> Can you rebase the PR. It looks like there were a few conflicts to be resolved

Done.

hashhar commented:
@cla-bot check

Praveen2112 commented:
@eformat Any updates on this PR ?

eformat commented:
> @eformat Any updates on this PR ?

hi, no. it is complete from a while back ? just waiting to  be merged really.

Praveen2112 commented:
Some of the review comments were not yet addressed. Can we address them and rebase the PR ?

eformat commented:
which comments please ?
also it has been a long time .. so i dont have a lot of bandwidth on this
at the moment
so if you do intend on merging it , please let me know as i do not have
bandwidth to continually rebase this over several months

On Thu, 2 Nov 2023 at 17:00, Praveen Krishna ***@***.***>
wrote:

> Some of the review comments were not yet addressed. Can we address them
> and rebase the PR ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/trinodb/trino/pull/17518#issuecomment-1790176290>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFN7IH25ZRPLMTRFUVW2WTYCNAJPAVCNFSM6AAAAAAYDNP4L6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTOOJQGE3TMMRZGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Praveen2112 commented:
Like this one - https://github.com/trinodb/trino/pull/17518#discussion_r1226218514, In this case can other members (or myself) work on top of your PR for applying the comments and rebasing, I will make sure you will be added as a co-author in the new PR.

eformat commented:
> Like this one - [#17518 (comment)](https://github.com/trinodb/trino/pull/17518#discussion_r1226218514), In this case can other members (or myself) work on top of your PR for applying the comments and rebasing, I will make sure you will be added as a co-author in the new PR.

this is endless. i did not see these on june 12 - sorry. sure - feel free to work on top of the existing code.

mosabua commented:
@Praveen2112 could you work with @eformat and continue this PR?

eformat commented:
FYI .. i am still keen to get this merged .. looks like its going to take a bit more effort .. i may have some bandwidth coming up to get that last mile in on this PR ....

eformat commented:
thats fine - @Praveen2112 as discussed - is that a clone of this PR ? if so i will close this and you can attribute as appropriate

ksobolew commented:
> is that a clone of this PR ?

No, it is an independent effort
